Menus/ObjectPickerGUI.cpp

```diff
@@ -25,1296 +12,613 @@
-    m_pGUIController->EnableMouse(pController->IsMouseControlled());
-
-    if (!s_pCursor)
-    {
-        ContentFile cursorFile("Base.rte/GUIs/Skins/Cursor.png");
-        s_pCursor = cursorFile.GetAsBitmap();
-    }
-
-    // Stretch the invisible root box to fill the screen
-	if (g_FrameMan.IsInMultiplayerMode())
-	{
-		dynamic_cast<GUICollectionBox *>(m_pGUIController->GetControl("base"))->SetSize(g_FrameMan.GetPlayerFrameBufferWidth(pController->GetPlayer()), g_FrameMan.GetPlayerFrameBufferHeight(pController->GetPlayer()));
+namespace RTE {
+
+	BITMAP *ObjectPickerGUI::s_Cursor = nullptr;
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	void ObjectPickerGUI::Clear() {
+		m_CursorPos.Reset();
+		m_GUIScreen = nullptr;
+		m_GUIInput = nullptr;
+		m_GUIControlManager = nullptr;
+		m_ParentBox = nullptr;
+		m_PopupBox = nullptr;
+		m_PopupText = nullptr;
+		m_GroupsList = nullptr;
+		m_ObjectsList = nullptr;
+		m_Controller = nullptr;
+		m_PickerState = PickerState::Disabled;
+		m_PickerFocus = PickerFocus::GroupList;
+		m_OpenCloseSpeed = 0.3F;
+		m_ModuleSpaceID = -1;
+		m_ShowType.clear();
+		m_NativeTechModule = 0;
+		m_ForeignCostMult = 4.0F;
+		m_SelectedGroupIndex = 0;
+		m_SelectedObjectIndex = 0;
+		m_PickedObject = nullptr;
+		m_RepeatStartTimer.Reset();
+		m_RepeatTimer.Reset();
+
+		m_ExpandedModules.resize(g_PresetMan.GetTotalModuleCount());
+		m_ExpandedModules.at(0) = true; // Base.rte is always expanded
+		for (int moduleID = 1; moduleID < m_ExpandedModules.size(); ++moduleID) {
+			m_ExpandedModules.at(moduleID) = false;
+		}
 	}
-	else
-	{
-		dynamic_cast<GUICollectionBox *>(m_pGUIController->GetControl("base"))->SetSize(g_FrameMan.GetResX(), g_FrameMan.GetResY());
+
+/////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////////
+
+	int ObjectPickerGUI::Create(Controller *controller, int whichModuleSpace, const std::string_view &onlyOfType) {
+		RTEAssert(controller, "No controller sent to ObjectPickerGUI on creation!");
+		m_Controller = controller;
+
+		if (!m_GUIScreen) { m_GUIScreen.reset(new AllegroScreen(g_FrameMan.GetBackBuffer8())); }
+		if (!m_GUIInput) { m_GUIInput.reset(new AllegroInput(controller->GetPlayer())); }
+		if (!m_GUIControlManager) { m_GUIControlManager.reset(new GUIControlManager()); }
+		if (!m_GUIControlManager->Create(m_GUIScreen.get(), m_GUIInput.get(), "Base.rte/GUIs/Skins/Base")) { RTEAbort("Failed to create GUI Control Manager and load it from Base.rte/GUIs/Skins/Base"); }
+
+		m_GUIControlManager->Load("Base.rte/GUIs/ObjectPickerGUI.ini");
+		m_GUIControlManager->EnableMouse(controller->IsMouseControlled());
+
+		if (!s_Cursor) {
+			ContentFile cursorFile("Base.rte/GUIs/Skins/Cursor.png");
+			s_Cursor = cursorFile.GetAsBitmap();
+		}
+
+		// Stretch the invisible root box to fill the screen
```

Comment:
So all the relative positioning is automagically adjusted to the actual screen size. There's no real point explaining it it's done pretty much everywhere.

Reply:
I'd thought, in a future comment, that it might be in order to allow the description text boxes to know how big the screen is so they don't go off the top of it.
Dunno if it's just for that purpose or also generally for positioning stuff.

Reply:
The root box is the "canvas" of the GUI elements that acts as the root parent.
It's created from ini with fixed dimensions that may not match the actual window dimensions, and that happens all the time unless you run 640x480.

Say your root box is 640x480 and the window dimensions are the same.
If you have an element that is anchored to the right edge of the screen and is supposed to be 100px off it, then everything is dandy.
But if your window dimensions are 740x480, that element is now 200px off the right edge, because the root box is still 640x480.
If you go ahead and manually reposition the box to be 100px off, it will clip outside the root box and won't be drawn (or part of it drawn if it's wide enough to go back into the root box area).
If you stretch the root box to 740x480 then the element will be adjusted to be 100px off the right edge like it is supposed to be without further input from you.

It's simple basic stuff that's more difficult to explain in words than in practice I guess.

Reply:
Ahh, and of course the gui is dumdum so you have to do this manually, instead of it automatically working this way. Cool, that all makes sense.

Well, given this is super common and basically necessary for the gui to work, I think you can kill off that comment.